### PR TITLE
Receive notifications for build failures

### PR DESCRIPTION
**Pull request by danidoni**
This PR allows subscribers to receive notifications when a package fails to build.

This is how a notification for a package build failure looks like:

image

This is a follow up of #13428 

Edit: Update screenshot

### CODE REVIEW

**Comment by dmarcoux** [comment]
It would be nice to have the factories for [notifications](https://github.com/openSUSE/open-build-service/blob/master/src/api/spec/factories/notification.rb) and [event subscriptions](https://github.com/openSUSE/open-build-service/blob/master/src/api/spec/factories/event_subscriptions.rb) regarding `Event::BuildFail`.

**Comment by danidoni** [comment]
> It would be nice to have the factories for [notifications](https://github.com/openSUSE/open-build-service/blob/master/src/api/spec/factories/notification.rb) and [event subscriptions](https://github.com/openSUSE/open-build-service/blob/master/src/api/spec/factories/event_subscriptions.rb) regarding `Event::BuildFail`.
> 
> This can happen in another PR.

I tried cramming that in this PR but it was too much. I left that for "later".

File: src/api/app/components/notification_notifiable_link_component.rb [hunk 1]
```diff
@@ -37,10 +37,17 @@ def notifiable_link_text
       else
         "Removed as #{role} of a project"
       end
+    when 'Event::BuildFail'
+      project = @notification.event_payload['project']
+      package = @notification.event_payload['package']
+      repository = @notification.event_payload['repository']
+      arch = @notification.event_payload['arch']
+      "Package #{package} on #{project} project failed to build against #{repository} / #{arch}"
```
Comment (dmarcoux) [inline src/api/app/components/notification_notifiable_link_component.rb:45]: This is hard to parse and distinguish notifications from one another, see this screenshot:

hard-to-parse

File: src/api/app/components/notification_component.rb [hunk 2]
```diff
@@ -13,7 +13,11 @@ def notification_icon
     when 'Comment'
       tag.i(class: ['fas', 'fa-comments'], title: 'Comment notification')
     else
-      tag.i(class: ['fas', 'fa-user-tag'], title: 'Relationship notification')
+      if @notification.event_type == 'Event::BuildFail'
+        tag.i(class: ['fas', 'fa-times-circle'], title: 'Package notification')
```
Comment (dmarcoux) [inline src/api/app/components/notification_component.rb:17]: We already use this icon for deleting content, so I suggest using another icon. The various icons for build statuses are defined in a helper:

https://github.com/openSUSE/open-build-service/blob/842c83e03dc0e48516d68595a296ae2dfae0ee0c/src/api/app/helpers/webui/buildresult_helper.rb#L2-L18

File: src/api/app/components/notification_action_description_component.rb [hunk 3]
```diff
@@ -28,6 +26,8 @@ def call
         "#{@user} made #{@recipient} #{@role} of #{@target_object}"
       when 'Event::RelationshipDelete'
         "#{@user} removed #{@recipient} as #{@role} of #{@target_object}"
+      when 'Event::BuildFail'
+        "Reason for the build: #{@notification.event_payload['reason']}"
```
[the remark this answers is not in the document]
Comment (danidoni) [reply src/api/app/components/notification_action_description_component.rb:30]: To understand why that build triggered

File: src/api/app/components/notification_notifiable_link_component.rb (hunk 1 above)
Comment (danidoni) [reply src/api/app/components/notification_notifiable_link_component.rb:45]: I think I'm going to improve over this in another PR because right now, I don't know what to add or remove to tell apart build fail event notifications from each other.

File: src/api/app/components/notification_action_description_component.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (hennevogel) [reply src/api/app/components/notification_action_description_component.rb:30]: And why is that important?

File: src/api/app/components/notification_notifiable_link_component.rb (hunk 1 above)
Comment (danidoni) [reply src/api/app/components/notification_notifiable_link_component.rb:45]: I created [a card](https://trello.com/c/wpOJunpB/118-improve-how-to-tell-apart-multiple-build-failure-event-notifications) to improve this later

File: src/api/app/components/notification_action_description_component.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmarcoux) [reply src/api/app/components/notification_action_description_component.rb:30]: So this was slightly changed in https://github.com/openSUSE/open-build-service/pull/13460/commits/8c7baec227dc45407cc12028b395af8f585bafbc.

**Review by dmarcoux — APPROVED** [review]
How notifications for a package build failure are displayed could be improved, but this is already a good first step. We also have this [card](https://trello.com/c/wpOJunpB/118-improve-how-to-tell-apart-multiple-build-failure-event-notifications) in our backlog for this.
